Framework/API/src/ExperimentInfo.cpp

```diff
@@ -335,29 +323,22 @@ void ExperimentInfo::setInstrument(const Instrument_const_sptr &instr) {
   invalidateAllSpectrumDefinitions();
 }
 
+Instrument_sptr ExperimentInfo::makeParameterizedInstrument() const {
+  populateIfNotLoaded();
+  checkDetectorInfoSize(*sptr_instrument, *m_detectorInfo);
+  return Geometry::ParComponentFactory::createInstrument(sptr_instrument,
+                                                         m_parmap);
+}
+
 /** Get a shared pointer to the parametrized instrument associated with this
 *workspace
 *
 *  @return The instrument class
 */
 Instrument_const_sptr ExperimentInfo::getInstrument() const {
-  populateIfNotLoaded();
-  checkDetectorInfoSize(*sptr_instrument, *m_detectorInfo);
-  auto instrument = Geometry::ParComponentFactory::createInstrument(
-      sptr_instrument, m_parmap);
-  // We can only set the DetectorInfo on the return Instrument if the API
-  // wrapper is fully constructed
-  if (m_detectorInfoWrapper) {
-    instrument->setDetectorInfo(m_detectorInfo,
-                                m_detectorInfoWrapper->detIdToIndexMap());
-  }
-  // We can only set the ComponentInfo on the return Instrument if the API
-  // wrapper is fully constructed
-  if (m_componentInfoWrapper) {
-    instrument->setComponentInfo(
-        m_componentInfo, m_componentInfoWrapper->componentIds(),
-        m_componentInfoWrapper->componentIdToIndexMap());
-  }
+  auto instrument = makeParameterizedInstrument();
+  instrument->setDetectorInfo(m_detectorInfo);
```

Comment:
Yes. I deliberately kept it out for the time being. But yes, all this information could be collected. Ideally, it would have just two construction related public functions (others are required due to `ComponentVisitor`).

```cpp
class InfoComponentVisitor : public ComponentVisitor {
public:
.....
std::unique<API::ComponentInfo> componentInfo() const;
std::unique<API::DetectorInfo> detectorinfo() const;
};
```